src/data/polynomial/reverse.lean

```diff
@@ -0,0 +1,206 @@
+begin
+  rw [rev_at, function.embedding.coe_fn_mk],
+  split_ifs with h j,
+  { exact nat.sub_sub_self h, },
+  { exact false.rec (N - i = i) (j (nat.sub_le N i)), },
+  { exact eq.refl i, },
+end
+
+@[simp] lemma rev_at_le {N i : ℕ} (H : i ≤ N) : rev_at N i = N - i :=
+if_pos H
+
+/-- `reflect N f` is the polynomial such that `(reflect N f).coeff i = f.coeff (rev_at N i)`.
+In other words, the terms with exponent `[0, ..., N]` now have exponent `[N, ..., 0]`.
+ In practice, `reflect` is only used when `N` is at least as large as the degree of `f`.
+ Eventually, it will be used with `N` exactly equal to the degree of `f`.  -/
+noncomputable def reflect (N : ℕ) (f : polynomial R) : polynomial R :=
+  finsupp.emb_domain (rev_at N) f
+
+lemma reflect_support (N : ℕ) (f : polynomial R) :
+  (reflect N f).support = image (rev_at N) f.support :=
+begin
+  ext1,
+  rw [reflect, mem_image, support_emb_domain, mem_map],
+end
+
+@[simp] lemma coeff_reflect (N : ℕ) (f : polynomial R) (i : ℕ) :
+  coeff (reflect N f) i = f.coeff (rev_at N i) :=
+calc finsupp.emb_domain (rev_at N) f i
+    = finsupp.emb_domain (rev_at N) f (rev_at N (rev_at N i)) : by rw rev_at_invol
+... = f.coeff (rev_at N i) : finsupp.emb_domain_apply _ _ _
+
+@[simp] lemma reflect_zero {N : ℕ} : reflect N (0 : polynomial R) = 0 := rfl
+
+@[simp] lemma reflect_eq_zero_iff {N : ℕ} {f : polynomial R} :
+  reflect N (f : polynomial R) = 0 ↔ f = 0 :=
+begin
+  split,
+  { intros a,
+    injection a with f0 f1,
+    rwa [map_eq_empty, support_eq_empty] at f0, },
+  { rintro rfl,
+    exact reflect_zero, },
+end
+
+@[simp] lemma reflect_add (f g : polynomial R) (N : ℕ) :
+  reflect N (f + g) = reflect N f + reflect N g :=
+by { ext1, rw [coeff_add, coeff_reflect, coeff_reflect, coeff_reflect, coeff_add], }
+
+@[simp] lemma reflect_C_mul (f : polynomial R) (r : R) (N : ℕ) :
+  reflect N (C r * f) = C r * (reflect N f) :=
+by { ext1, rw [coeff_reflect, coeff_C_mul, coeff_C_mul, coeff_reflect], }
+
+@[simp] lemma reflect_C_mul_X_pow (N n : ℕ) {c : R} :
+  reflect N (C c * X ^ n) = C c * X ^ (rev_at N n) :=
+begin
+  ext,
+  rw [reflect_C_mul, coeff_C_mul, coeff_C_mul, coeff_X_pow, coeff_reflect],
+  split_ifs,
+  { rw [h, rev_at_invol, coeff_X_pow_self], },
+  { rw [not_mem_support_iff_coeff_zero.mp],
+    intro a,
+    rw [← one_mul (X ^ n), ← C_1] at a,
+    apply h,
+    rw [← (mem_support_C_mul_X_pow a), rev_at_invol], },
+end
+
+@[simp] lemma reflect_monomial (N n : ℕ) : reflect N ((X : polynomial R) ^ n) = X ^ (rev_at N n) :=
+by rw [← one_mul (X ^ n), ← one_mul (X ^ (rev_at N n)), ← C_1, reflect_C_mul_X_pow]
+
+lemma reflect_mul_induction (cf cg : ℕ) :
+ ∀ N O : ℕ, ∀ f g : polynomial R,
+ f.support.card ≤ cf.succ → g.support.card ≤ cg.succ → f.nat_degree ≤ N → g.nat_degree ≤ O →
+ (reflect (N + O) (f * g)) = (reflect N f) * (reflect O g) :=
+begin
+  induction cf with cf hcf,
+  --first induction: base case
+  { induction cg with cg hcg,
```

Comment:
I suspect it might be easier to `import data.polynomial.div` and use `refine f.rec_on_horner _ _ _` than do induction on the cardinality of the support. But I didn't try it out in much detail.

Reply:
Thank you for all the remaining suggestions: I liked them all!

I will try to simplify further the induction and look into ```rec_on_horner```.  I saw that it uses ```Sort```s, which means that it might take me some time to parse it.  Anyway, thanks a lot!

Reply:
You're welcome, thanks for doing the PR!

`Sort`s are not too hard if you know what `Type u` means. `Sort 0` is `Prop` and `Sort (u + 1)` is `Type u`. Writing `Sort u` instead of `Type*` is a way to make definitions a bit more generic, allowing `Prop` and `Type` without re-defining anything. (In fact it's the other way around: `Prop` is defined as an abbreviation of `Sort 0` and `Type u` as an abbreviation of `Sort (u + 1)`.)

Reply:
Thanks for the explanation!  In the meantime, I just started playing around with the code that you suggested and it works **very** well!  Here is what the proof looks like, right now!

```
lemma reflect_mul_induction2 (cf cg : ℕ) :
 ∀ N O : ℕ, ∀ f g : polynomial R,
 f.support.card ≤ cf.succ → g.support.card ≤ cg.succ → f.nat_degree ≤ N → g.nat_degree ≤ O →
 (reflect (N + O) (f * g)) = (reflect N f) * (reflect O g) :=
begin
  intros N O f g cf cg,
  refine f.rec_on_horner _ _ _,
  { simp only [reflect_zero, zero_mul, eq_self_iff_true, forall_true_iff], },
  { intros df dg cp0 cg cg0 hg ec0,
    simp only [*, reflect_mul] at *, },
  { intros p p0 hh pX g0,
    rw reflect_mul pX g0, },
end
```

Reply:
I am in the process of cleaning up unnecessary hypotheses, while understanding the mechanics of this induction.

Reply:
Sorry, I messed up: the very short proof above *uses* ```reflect_mul```, since I was doing something silly.  I will try to see how it would work without tautologically proving it from itself!

Reply:
@Vierkantor , I have been trying to use ```rec_on_horner``` and it seems that it invariably wants me to prove the ```reflect_mul``` in the special case in which one of the two factors is a monomial.  This is of course true, provable (I had it in an earlier version), but not necessarily shorter than the current induction proof, I think.  (The special case involves understanding the relation between reflecting and multiplying by a monomial, that is, translating the support.)

Unless you have an idea on how to simplify this further, I am not sure that I prefer to use ```rec_on_horner```.  What do you think?

Reply:
Hmm, then induction on the support sounds like the best idea. Let me see if I can do some lower-level refactoring to clean up the proof...

Reply:
That would be great: I will look carefully at the diff afterwards, to learn how to work with Lean better!

Reply:
If we add the following helper lemma:

```lean
lemma rev_at_add {N O n o : ℕ} (hn : n ≤ N) (ho : o ≤ O) :
  rev_at (N + O) (n + o) = rev_at N n + rev_at O o :=
begin
  rcases nat.le.dest hn with ⟨n', rfl⟩,
  rcases nat.le.dest ho with ⟨o', rfl⟩,
  repeat { rw rev_at_le },
  { rw [add_assoc, add_left_comm n' o, ← add_assoc], simp },
  { exact le_add_right (le_refl _), },
  { exact le_add_right (le_refl _), },
  { exact add_le_add hn ho }
end
```

Then I can shorten it to:
```lean
begin
  induction cf with cf hcf,
  --first induction: base case
  { induction cg with cg hcg,
    -- second induction: base case
    { intros N O f g Cf Cg Nf Og,
      rw [C_mul_X_pow_of_card_support_le_one Cf, C_mul_X_pow_of_card_support_le_one Cg],
      simp only [mul_assoc, X_pow_mul, ← pow_add X, reflect_C_mul, reflect_monomial,
                 add_comm, rev_at_add Nf Og] },
    -- second induction: induction step
    { intros N O f g Cf Cg Nf Og,
      by_cases g0 : g = 0,
      { simp [g0] },

      rw [← erase_lead_add_C_mul_X_pow g, mul_add, reflect_add, reflect_add, mul_add, hcg, hcg];
        try { assumption },
      { exact le_add_left card_support_C_mul_X_pow_le_one },
      { exact (le_trans (nat_degree_C_mul_X_pow_le g.leading_coeff g.nat_degree) Og) },
      { exact nat.lt_succ_iff.mp (gt_of_ge_of_gt Cg (erase_lead_support_card_lt g0)) },
      { exact le_trans erase_lead_nat_degree_le Og } } },
  --first induction: induction step
  { intros N O f g Cf Cg Nf Og,
    by_cases f0 : f=0,
    { simp [f0] },
    { rw [← erase_lead_add_C_mul_X_pow f, add_mul, reflect_add, reflect_add, add_mul, hcf, hcf];
         try { assumption },
      { exact le_add_left card_support_C_mul_X_pow_le_one },
      { exact (le_trans (nat_degree_C_mul_X_pow_le f.leading_coeff f.nat_degree) Nf) },
      { exact nat.lt_succ_iff.mp (gt_of_ge_of_gt Cf (erase_lead_support_card_lt f0)) },
      { exact (le_trans erase_lead_nat_degree_le Nf) } } },
end
```

Reply:
Thanks a lot!  I like your suggestion and I have just pushed it!